src/main/java/appeng/thirdparty/codechicken/lib/model/Quad.java

```diff
@@ -0,0 +1,557 @@
+		 */
+		public void preProcess()
+		{
+			if( format.hasPosition )
+			{
+				vec = raw[format.positionIndex];
+			}
+			if( format.hasNormal )
+			{
+				normal = raw[format.normalIndex];
+			}
+			if( format.hasColor )
+			{
+				color = raw[format.colorIndex];
+			}
+			if( format.hasUV )
+			{
+				uv = raw[format.uvIndex];
+			}
+			if( format.hasLightMap )
+			{
+				lightmap = raw[format.lightMapIndex];
+			}
+		}
+
+		/**
+		 * Gets the 2d X coord for the given axis.
+		 *
+		 * @param s The axis. side >> 1
+		 *
+		 * @return The x coord.
+		 */
+		public float dx( int s )
+		{
+			if( s <= 1 )
+			{
+				return vec[0];
+			}
+			else
+			{
+				return vec[2];
+			}
+		}
+
+		/**
+		 * Gets the 2d Y coord for the given axis.
+		 *
+		 * @param s The axis. side >> 1
+		 *
+		 * @return The y coord.
+		 */
+		public float dy( int s )
+		{
+			if( s > 0 )
+			{
+				return vec[1];
+			}
+			else
+			{
+				return vec[2];
+			}
+		}
+
+		/**
+		 * Interpolates the new color values for this Vertex using the others as a reference.
+		 *
+		 * @param interpHelper The InterpHelper to use.
+		 * @param others The other Vertices to use as a template.
+		 *
+		 * @return The same Vertex.
+		 */
+		public Vertex interpColorFrom( InterpHelper interpHelper, Vertex[] others )
+		{
+			for( int e = 0; e < 4; e++ )
+			{
+				float p1 = others[0].color[e];
+				float p2 = others[1].color[e];
+				float p3 = others[2].color[e];
+				float p4 = others[3].color[e];
+				//Only interpolate if colors are different.
+				if( p1 != p2 || p2 != p3 || p3 != p4 )
+				{
+					color[e] = interpHelper.interpolate( p1, p2, p3, p4 );
+				}
+			}
+			return this;
+		}
+
+		/**
+		 * Interpolates the new UV values for this Vertex using the others as a reference.
+		 *
+		 * @param interpHelper The InterpHelper to use.
+		 * @param others The other Vertices to use as a template.
+		 *
+		 * @return The same Vertex.
+		 */
+		public Vertex interpUVFrom( InterpHelper interpHelper, Vertex[] others )
+		{
+			for( int e = 0; e < 2; e++ )
+			{
+				float p1 = others[0].uv[e];
+				float p2 = others[1].uv[e];
+				float p3 = others[2].uv[e];
+				float p4 = others[3].uv[e];
+				if( p1 != p2 || p2 != p3 || p3 != p4 )
+				{
+					uv[e] = interpHelper.interpolate( p1, p2, p3, p4 );
+				}
+			}
+			return this;
+		}
+
+		/**
+		 * Interpolates the new LightMap values for this Vertex using the others as a reference.
+		 *
+		 * @param interpHelper The InterpHelper to use.
+		 * @param others The other Vertices to use as a template.
+		 *
+		 * @return The same Vertex.
+		 */
+		public Vertex interpLightMapFrom( InterpHelper interpHelper, Vertex[] others )
+		{
+			for( int e = 0; e < 2; e++ )
+			{
+				float p1 = others[0].lightmap[e];
+				float p2 = others[1].lightmap[e];
+				float p3 = others[2].lightmap[e];
+				float p4 = others[3].lightmap[e];
+				if( p1 != p2 || p2 != p3 || p3 != p4 )
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Equality tests should not be made with floating point values. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1244)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Equality tests should not be made with floating point values. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1244)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Equality tests should not be made with floating point values. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1244)